src/core/packages/http/rate-limiter-internal/src/service.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import {
+  BehaviorSubject,
+  endWith,
+  map,
+  skipUntil,
+  type Observable,
+  Subject,
+  takeUntil,
+} from 'rxjs';
+import type { CoreService } from '@kbn/core-base-server-internal';
+import type { KibanaRequest, OnPreAuthHandler } from '@kbn/core-http-server';
+import type { InternalHttpServiceSetup } from '@kbn/core-http-server-internal';
+import type { EluMetrics } from '@kbn/core-metrics-server';
+import type { InternalMetricsServiceSetup } from '@kbn/core-metrics-server-internal';
+
+/** @internal */
+export interface SetupDeps {
+  http: InternalHttpServiceSetup;
+  metrics: InternalMetricsServiceSetup;
+}
+
+/** @internal */
+export type InternalRateLimiterSetup = void;
+
+/** @internal */
+export type InternalRateLimiterStart = void;
+
+/** @internal */
+export class HttpRateLimiterService
+  implements CoreService<InternalRateLimiterSetup, InternalRateLimiterStart>
+{
+  private overloaded$ = new BehaviorSubject(false);
+  private ready$ = new Subject<boolean>();
+  private stopped$ = new Subject<boolean>();
+
+  private handler: OnPreAuthHandler = (request, response, toolkit) => {
+    if (!this.shouldBeThrottled(request)) {
+      return toolkit.next();
+    }
+
+    return response.customError({
```

Comment:
Wonder if we could add some logging here - I just noticed this PR from the weekly AppEx update for 2025-02-20.

I worried that we may not have easy visibility that this occurring; perhaps it is logged elsewhere along the code path.  But if it isn't, and someone is just looking at Kibana logs for problem diagnosis, how would they figure out this was happening?

You could look in the Kibana proxy logs (for instance, in a cloud env), but this is something we rarely have to do, I don't think folks would just happen to look through these.  So I'm thinking this won't be noticed, at least via just looking at the Kibana logs.

If we do this, would probably be good to add some kind of unique tag that would be easy to search for: `throttled-429` or such ...

Reply:
cleaning up my browser tabs, this was still open, so I created issue https://github.com/elastic/kibana/issues/212814 to track